DESIGN.md

```diff
@@ -0,0 +1,756 @@
+completion queues and a Tokio current-thread epoll-based runtime. Instead of
+waiting on completion events by blocking the thread on the io-uring completion
+queue, the tokio-uring runtime registers the completion queue with the epoll
+handle. By building tokio-uring on top of Tokio's runtime, existing Tokio
+ecosystem crates can work with the tokio-uring runtime.
+
+When the kernel pushes a completion event onto the completion queue,
+"epoll_wait" unblocks and returns a readiness event. The Tokio current-thread
+runtime then polls the io-uring driver task, draining the completion queue and
+notifying completion futures. 
+
+Like Tokio, using an I/O type does not require explicitly referencing the
+runtime. Operations access the current runtime via a thread-local variable.
+Requiring a handle would be intrusive as the application must pass the handle
+throughout the code. Additionally, an explicit handle would require a
+pointer-sized struct, causing binary bloat. The disadvantage with such an
+approach is that there is no way to guarantee operation submission happens
+within the runtime context at compile time. Attempting to use a tokio-uring
+resource from outside of the runtime will result in a panic.
+
+```rust
+use tokio_uring::runtime::Runtime;
+use tokio_uring::net::TcpListener;
+
+fn main() {
+    // Binding a TcpListener does not require access to the runtime.
+    let listener = TcpListener::bind("0.0.0.0:1234".parse().unwrap());
+
+    let rt = Runtime::new().unwrap();
+    rt.block_on(async {
+        // This works, as `block_on` sets the thread-local variable.
+        let _ = listener.accept().await;
+    });
+
+    // BOOM: panics because called outside of the runtime
+    futures::future::block_on(async {
+        let _ = listener.accept().await;
+    });
+}
+```
+
+## Operation state
+
+Most io-uring operations reference resources, such as buffers and file
+descriptors, for the kernel to use. These resources must remain available while
+the operation is in-flight. Any memory referenced by pointers must remain
+allocated, and the process must not access the memory. Because asynchronous Rust
+allows dropping futures at any time, the operation futures may not own data
+referenced by the in-flight operation. The tokio-uring runtime will take
+ownership and store resources referenced by operations while they are in-flight.
+
+```rust
+struct IoUringDriver {
+    // Storage for state referenced by in-flight operations
+    in_flight_operations: Slab<Operation>,
+    
+    // The io-uring submission and completion queues.
+    queues: IoUringQueues,
+}
+
+struct Operation {
+    // Resources referenced by the kernel, this must stay
+    // available until the operation completes.
+    state: State,
+    lifecycle: Lifecycle,
+}
+
+enum State {
+    // An in-flight read operation, `None` when reading into
+    // a kernel-owned buffer pool
+    Read { buffer: Option<buf::Slice> },
```

Comment:
Is it better to have two different State variations where we have a generic for kernel-owned buffer pool and Vec (if they are always the same)? Will one state be `None` and another state be `Some` for `Read`?

Reply:
It might be, but before making that tweak I would like to measure the impact. It doesn't materially change the design though so we can explore it in the future.